lib/rewrite-pages.js

```diff
@@ -0,0 +1,214 @@
+      continue;
+    }
+
+    break;
+  }
+
+  toRemove.forEach(e => e.remove());
+
+  a.href = protocols.rewrite(url);
+
+  // trailing non-link text
+  frag.appendChild(doc.createTextNode(text.substring(lastMatch)));
+  parent.replaceChild(frag, toReplace);
+
+  return a;
+}
+
+function processElement(element) {
+  if (element.namespaceURI != "http://www.w3.org/1999/xhtml")
+    return;
+  if (element.localName == "a" && element.hasAttribute("href")) {
+    // .href resolves relative to base domain, .getAttribute may return relative urls e.g. /ipfs/...
+    let orig = element.getAttribute("href");
+    let rewritten = protocols.rewrite(orig);
+    if (rewritten && orig != rewritten) {
+      element.href = rewritten;
+    }
+
+  }
+}
+
+function mutationCallback(records) {
+  for (let r of records) {
+    if (r.type == "childList") {
+      for (let added of r.addedNodes) {
+
+        if (added.nodeType == 3) { // Node.TEXT_NODE
+          //console.log("added T", added.localname)
+          processTextNode(added);
+        }
+
+        if (added.nodeType == 1) { // Node.ELEMENT
+          //console.log("added E", added)
+          processElement(added);
+        }
+      }
+    }
+    if (r.type == "characterData") {
+      //console.log("changed T", r.target)
+      processTextNode(r.target);
+    }
+  }
+}
+
+const SCHEME_WHITELIST = "resource http https ftp ipfs fs ipns".split(" ");
+
+
+const documentObserver = function(event) {
+  let { subject, type } = event;
+
+  if (type != 'content-document-global-created')
+    return;
+  if (!gw.linkify)
+    return;
+  if (!(subject instanceof Ci.nsIDOMWindow))
+    return;
+
+  let window = subject;
+  let document = window.document;
+
+  let uri = document.documentURIObject;
+
+  //don't touch chrome/about documents
+  // alternative approach: check for non-system principal
+  if (!uri || SCHEME_WHITELIST.indexOf(uri.scheme) === -1)
+    return;
+
+
+  /*
+  // approach A
+  // tree walk on load -> might lead to a long processing pause on large docs
+
+  let docLoaded = function() {
+    let obs = new window.MutationObserver(mutationCallback)
+    obs.observe(document, {characterData: true, subtree: true, childList: true});
+    let walker = document.createTreeWalker(document.documentElement, window.NodeFilter.SHOW_TEXT);
+    let next;
+    while(walker.nextNode()) {
+      let replaced = processTextNode(walker.currentNode);
+      if(replaced) {
+        walker.currentNode = replaced;
+      }
+    }
+    document.removeEventListener("DOMContentLoaded", docLoaded);
+  }
+
+  document.addEventListener("DOMContentLoaded", docLoaded);
+  */
+
+  // approach B
+  // incremental, process text as it arrives from the parser
+  let obs = new window.MutationObserver(mutationCallback);
+  obs.observe(document, {
+    characterData: true,
+    subtree: true,
+    childList: true
+  });
+
+};
+
+events.on('content-document-global-created', documentObserver);
```

Comment:
It may be a better idea to on/off entire listener based on checkbox state.
Perhaps something along these lines:

```javascript
gw.onChange(() => {
  if (gw.linkify) {
    events.on('content-document-global-created', documentObserver);
  } else {
    events.off('content-document-global-created', documentObserver);
  }
});
gw.reload();
```